Source/DafnyLanguageServer/Language/Symbols/DafnyLangSymbolResolver.cs

```diff
@@ -150,9 +154,60 @@ private class SymbolDeclarationResolver {
           cancellationToken.ThrowIfCancellationRequested();
           functionSymbol.Parameters.Add(ProcessFormal(scope, parameter));
         }
+        if (function.Result != null) {
+          functionSymbol.Parameters.Add(ProcessFormal(scope, function.Result));
+        }
+
+        Func<Expression, ScopeSymbol?> subP = (Expression e) => ProcessFunctionExpr(functionSymbol, e);
+        functionSymbol.Body = ProcessFunctionExpr(functionSymbol, function.Body);
+        functionSymbol.Ens = ProcessListAttributedExpressions(function.Ens, subP);
+        functionSymbol.Req = ProcessListAttributedExpressions(function.Req, subP);
+        functionSymbol.Reads = ProcessListFramedExpressions(function.Reads, subP);
+        functionSymbol.Decreases = ProcessListExpressions(function.Decreases.Expressions, subP);
+        functionSymbol.ByMethodBody = ProcessFunctionByMethodBody(functionSymbol, function.ByMethodBody);
```

Comment:
All these properties might be irrelevant for the hover feature. Or do you have an example?

Reply:
Yes:
```
  predicate Sorted(a: array<O.T>, low: int, high: int)
    requires 0 <= low <= high <= a.Length
    reads a
    // The body of the predicate is hidden outside the module, but the postcondition is
    // "exported" (that is, visible, known) outside the module.  Here, we choose the
    // export the following property:
    ensures Sorted(a, low, high) ==> forall i, j :: low <= i < j < high ==> O.Leq(a[i], a[j])
  {
    forall i, j :: low <= i < j < high ==> O.Leq(a[i], a[j])
  }
  ```
  
In the "ensures" clause, there is a forall now introducing variables, which is accounted as a new scope and scope-bearing expression node. Without the change you commented, there wasn't any scope associated to the ForAll node and thus it was crashing on the file `SymbolTableFactory` on the line `currentScope = declarations[node];`

The rationale is, now that we have scopes for expressions and not just for statements, we need to create specific sub-scopes for each expression that methods and functions rely on, because each of them might introduce variables, like in comprehensions or let expressions.: